### PR TITLE
Fix #3477: Add custom classpath to dotr script

### LINKED ISSUE #3477: `dist/bin/dotr` can't run with a custom `-classpath`
```
sbt dist-bootstrapped/pack
echo "object Main { def main(a: Array[String]): Unit = println(\"OK\") }" > Main.scala
./dist-bootstrapped/target/pack/bin/dotc -d out Main.scala
./dist-bootstrapped/target/pack/bin/dotr -classpath out Main
```

The last line will default to the repl instead running the application.

Note that `dist-bootstrapped/target/pack/bin/dotr` is generated from `dist/bin/dotr` when packing.

### PR DESCRIPTION
**Pull request by rsoeldner**
@nicolasstucki This addresses issue #3477

### CODE REVIEW

File: dist/bin/dotr [hunk 1]
```diff
@@ -31,18 +31,39 @@ source "$PROG_HOME/bin/common"
...
+run_app=flase
+while [[ $# -gt 0 ]]; do
+key=$1
+case $key in
+        -repl)
+         run_repl=true
+         shift
+         ;;
+        -classpath)
+         CLASS_PATH+=":$2"
```
Comment (nicolasstucki) [inline dist/bin/dotr]: We should probably add the user-defined classpaths to the left of the classpath to give them the priority.

File: dist/bin/dotr [hunk 2]
```diff
@@ -31,18 +31,39 @@ source "$PROG_HOME/bin/common"
...
+          shift
+          shift
+          ;;
+          *)
+           addResidual "$1"
+           shift
+           ;;
+        -d)
+        DEBUG="$DEBUG_STR"
```
Comment (nicolasstucki) [inline dist/bin/dotr]: Is this ever reached with the `*)` before it?

File: dist/bin/dotr [hunk 3]
```diff
@@ -31,18 +31,39 @@ source "$PROG_HOME/bin/common"
...
+         shift
+         ;;
+  esac
+done
+
+if [ -z $residual_args ]; then
     echo "Starting dotty REPL..."
     eval "$PROG_HOME/bin/dotc -repl"
-elif [[ ${first_arg:0:1} == "-" ]]; then
-    eval "$PROG_HOME/bin/dotc -repl $@"
```
Comment (nicolasstucki) [inline dist/bin/dotr]: We should still be able to call the repl with a custom classpath and custom args. You can assume that the `-repl` is set in that case and hence `run_repl` should be true.

**Comment by nicolasstucki** [comment]
You could also enable the two tests in https://github.com/lampepfl/dotty/blob/master/project/scripts/sbtBootstrappedTests#L29. That way we can avoid some regressions

**Comment by nicolasstucki** [comment]
Otherwise the script code LGTM

File: dist/bin/dotr [hunk 4]
```diff
@@ -31,18 +31,49 @@ source "$PROG_HOME/bin/common"
...
+           ;;
+
+  esac
+done
+
+CLASS_PATH="${classpath_args[@]}$CLASS_PATH"
+
+if [ $run_repl == true ]; then
+    echo "Starting dotty REPL"
+    echo "$PROG_HOME/bin/dotc -classpath $CLASS_PATH -repl ${residual_args[@]}"
```
Comment (nicolasstucki) [inline dist/bin/dotr]: Here the classpath should only contain `classpath_args` if it is not empty.

What does it mean exactly `classpath_args[@]`? I'm not familiar with the `[@]` part.

Comment (rsoeldner) [reply dist/bin/dotr]: @nicolasstucki Afaik, the `[*]` or `[@]` subscripts differ only within double quotes, with `[*]` each element inside the array is seperated by `IFS`

**Comment by rsoeldner** [comment]
@nicolasstucki If everything is ok, I would rebase to a single commit.

**Comment by allanrenucci** [comment]
I’d like to review this before we merge

**Comment by allanrenucci** [comment]
@nicolasstucki can you review the last commit

**Pull request merged**